### DISCUSSION TITLE
校园网下无法获取IPv6公网地址

### DISCUSSION THREAD

**Discussion by CAB233** [immortalwrt/immortalwrt#1239] [Q&A] [answered]
路由器使用23.05分支自己编译的包，只添加了SmartDNS用于分流校园网内外网。

使用电脑进行拨号时可以获得一个2001前缀的v6地址，可以ping通外布v6地址。

而使用路由器进行拨号时，无论是使用默认配置（只在wan里填写PPPoE账号）还是根据[这个教程](https://sustech.online/service/network/ipv6/)都只能获取fe80这个v6地址，路由器自身也无法ping通外部v6地址
![1](https://github.com/immortalwrt/immortalwrt/assets/69302204/16060e47-f450-4fce-ac9f-5e6df57cb9ca)

根据 https://github.com/immortalwrt/immortalwrt/discussions/773#discussioncomment-3568338 提到的ipv6helper，我看已经默认编译在了固件里。

所以请教一下要如何配置先让路由器能够获取2001的v6

**Comment by science2468** [comment]
把wan6删了。pppoe会自动生成一个不可编辑的虚拟接口。

**Reply by CAB233** [reply]
这个「不可编辑的虚拟接口」是图里的虚拟动态接口吗，删的前后v6没什么变化，路由器依然ping不通v6地址
![图片](https://github.com/immortalwrt/immortalwrt/assets/69302204/19b94f14-4755-4307-bd3d-66d01496b7f5)

**Reply by science2468** [reply]
截图就这么点？保存并应用后重启pppoe

**Reply by CAB233** [reply]
是的，截图里就是全部设备了，pppoe重启或者设备重启都是一样只能获取fe80的地址

**Reply by science2468** [reply]
啊？要不你给个向日葵识别码，让我瞧瞧？

**Reply by science2468** [reply]
![image](https://github.com/immortalwrt/immortalwrt/assets/102643084/b05637aa-29be-4fef-a6b8-a6494d2c52d4)
成功的话就是图片这样。要是不行换换固件试试，要不直接不用ipv6。

**Comment by 1715173329** [comment]
校园网是世纪难题，再加上 odhcpd 也挺烂的。可以抓包看看。

**Reply by CAB233** [reply]
我不清楚抓包，需要获取哪些信息呢

**Comment by CAB233** [comment]
目前另刷xwrt和官方openwrt稳定版并没有改变情况，另外在路由器日志里发现`daemon.warn odhcp6c[23708]: Server returned IA_PD status 'No Prefix Available (NoPrefixAvail)'`刷屏

**Reply by science2468** [reply]
你们校园网不给ipv6前缀。要不不用ipv6；要不去问卖你校园网的。

**Comment by CAB233** [comment]
抓包后发现校园网确实分配了公网v6地址，所以问题应该在路由器的设置上。

借了个TPLINK的路由器做了测试，在设置里关闭「前缀授权」，WAN口就能拿到公网v6地址。

然后我尝试在immortalwrt上关闭前缀授权的类似选项
![2](https://github.com/immortalwrt/immortalwrt/assets/69302204/99facd67-29cb-4230-91fd-d6dbdb888db1)

保存上述配置重启路由器，但我在查看进程时发现仍带有`-P0`参数
![1](https://github.com/immortalwrt/immortalwrt/assets/69302204/21c884e5-3329-4a2c-bc01-6ef08ead1fdd)

所以我该如何正确关闭WAN口的前缀授权？

**Reply by science2468** [reply]
把lan口的也关了试试。

**Reply by CAB233** [reply]
没有效果，进程里也还有P0参数

**Reply by 1715173329** [reply]
手动改 network 的 wan 配置，增加 `option reqprefix 'no'`

**Comment by CAB233** [comment, marked as answer]
非常感谢各位，目前已经成功拿到公网v6了，校园网（或者说教育网更准确，运营商的网络基本即插即用）确实挺神奇，很多教程的起始点几乎一致但最后实现细节却各有不同，因此在这记录一下给后人参考。

我这里的校园网也是不会下发前缀而只会下发地址，类似2001:xxxx:xxxx/128。只在wan口进行正常拨号，之后我的界面大概是这样
![](https://github.com/immortalwrt/immortalwrt/assets/69302204/921ec370-3e02-4a0a-b8d7-aa3da20953c0)

路由器要获取公网v6,大概需要这些设置即可
![](https://github.com/immortalwrt/immortalwrt/assets/69302204/de0dbf3e-0e7b-4cb0-9f78-d3d6e71b76f4)
![](https://github.com/immortalwrt/immortalwrt/assets/69302204/6e700499-f576-47a7-bb5c-548b8984c83b)

成功后界面如下
![](https://github.com/immortalwrt/immortalwrt/assets/69302204/ccc6bc70-d114-40f4-8297-30fd41fc4131)

------分割线------

其中关于wan6的「设备」配置有一点小乌龙。在进行拨号后wan的「设备」会变成pppoe-wan,而在配置文件`/etc/config/network`里wan的「设备」仍为wan（重启路由器后仍是这样）。

根据[openwrt wiki中关于wan6的配置](https://openwrt.org/docs/guide-user/network/ipv6/configuration#native_ipv6_connection)提到use same device as in wan-section or "@wan"，而此时luci中设备名不一致但配置文件中却是一致的！不过这里也是小问题，可以在luci或配置文件中手动配置wan6的设备为@wan或pppoe-wan，重启一下设备后就能在wan6看到公网v6地址了。
![](https://github.com/immortalwrt/immortalwrt/assets/69302204/eab97fd0-66a1-47c4-ac4b-b7a3ae148e9b)
